src/Polysemy/Input/Streaming.hs

```diff
@@ -0,0 +1,85 @@
+module Polysemy.Input.Streaming
+  ( -- * Underlying Effect
+    module Polysemy.Input
+
+    -- * Actions
+  , yieldInput
+  , yieldRace
+  , exhaust
+
+    -- * Intepretations
+  , runInputViaStream
+  , runInputViaInfiniteStream
+  ) where
+
+import qualified Control.Concurrent.Async as A
+import           Data.Functor.Of
+import           Polysemy
+import           Polysemy.Final
+import           Polysemy.Input
+import           Polysemy.State
+import qualified Streaming as S
+import qualified Streaming.Prelude as S
+
+
+
+runInputViaStream
+    :: S.Stream (Of i) (Sem r) ()
+    -> InterpreterFor (Input (Maybe i)) r
+runInputViaStream stream
+  = evalState (Just stream)
```

Comment:
Why wrapping the state in a Just? wouldn't below have the same semantics?

```
runInputViaStream
    :: forall i r a.S.Stream (Of i) (Sem r) ()
    -> Sem (Input (Maybe i) ': r) a -> Sem r a
runInputViaStream stream
  = evalState stream
  . reinterpret ( \Input -> do
      s <- get @(S.Stream (Of i) (Sem r) ())
      raise (S.uncons s) >>= \case
        Nothing -> pure Nothing
        Just (i, s') -> do
          put s'
          pure $ Just i
  )

```

Reply:
This will duplicate the actions at the end of the computation --- consider an action that only does effects and never `yield`s:

```haskell
foo :: Stream (Of i) IO a
foo = liftIO $ putStrLn "hello"
```

under the `runInputViaStream` above, this will print `hello` for every call to `input`!

Reply:
Aha, make sense. then shouldn't Left () case have 'put Nothing'?